src/Lean/Meta/Eval.lean

```diff
@@ -4,19 +4,21 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Sebastian Ullrich, Leonardo de Moura
 -/
 import Lean.Meta.Check
+import Lean.Util.CollectLevelParams
 
 namespace Lean.Meta
 
 unsafe def evalExprCore (α) (value : Expr) (checkType : Expr → MetaM Unit) (safety := DefinitionSafety.safe) : MetaM α :=
   withoutModifyingEnv do
     let name ← mkFreshUserName `_tmp
     let value ← instantiateMVars value
+    let us := collectLevelParams {} value |>.params
```

Comment:
`elabEvalUnsafe` uses

https://github.com/leanprover/lean4/blob/8335d4e27d7a44ee13ad46d9aaa33c78b1984183/src/Lean/Elab/BuiltinCommand.lean#L634-L637

which rather puzzlingly uses `instantiateMVars` twice. I don't really know which if either is right, so I've limited this to the minimal change that fixes my bug.

Reply:
@kmill, do you have any guesses as to why `inferType` happens to the pre-`instantiateMVars`d value in the other file referenced above?

Reply:
@digama0, @semorrison, do you have any thoughts on what's going on here?

Reply:
I don't see a reason to `instantiateMVars` that second time, but I do wonder if it should be `let type ← instantiateMVars type` instead, since there's no reason why the inferred type will be metavariable-free.

Reply:
How can a value have no metavariables but its inferred type have some?

Reply:
If the value is `a` where `a : ?m1` is a local constant then this can happen.

Reply:
But `eval` can't work if there are free variables anyway, so presumably this doesn't matter?